pkg/components/pubsub/pluggable.go

```diff
@@ -0,0 +1,210 @@
+// It also fetches and set the component features.
+func (p *grpcPubSub) Init(metadata pubsub.Metadata) error {
+	if err := p.Dial(metadata.Name); err != nil {
+		return err
+	}
+
+	protoMetadata := &proto.MetadataRequest{
+		Properties: metadata.Properties,
+	}
+
+	// TODO Static data could be retrieved in another way, a necessary discussion should start soon.
+	// we need to call the method here because features could return an error and the features interface doesn't support errors
+	featureResponse, err := p.Client.Features(p.Context, &proto.FeaturesRequest{})
+	if err != nil {
+		return err
+	}
+
+	p.features = make([]pubsub.Feature, len(featureResponse.Feature))
+	for idx, f := range featureResponse.Feature {
+		p.features[idx] = pubsub.Feature(f)
+	}
+
+	_, err = p.Client.Init(p.Context, &proto.PubSubInitRequest{
+		Metadata: protoMetadata,
+	})
+
+	return err
+}
+
+// Features lists all implemented features.
+func (p *grpcPubSub) Features() []pubsub.Feature {
+	return p.features
+}
+
+// Publish publishes data to a topic.
+func (p *grpcPubSub) Publish(req *pubsub.PublishRequest) error {
+	_, err := p.Client.Publish(p.Context, &proto.PublishRequest{
+		Topic:      req.Topic,
+		PubsubName: req.PubsubName,
+		Data:       req.Data,
+		Metadata:   req.Metadata,
+	})
+	return err
+}
+
+type messageHandler = func(*proto.PullMessageResponse)
+
+// adaptHandler returns a non-error function that handle the message with the given handler and ack when returns.
+//
+//nolint:nosnakecase
+func (p *grpcPubSub) adaptHandler(ctx context.Context, streamingPull proto.PubSub_PullMessagesClient, handler pubsub.Handler) messageHandler {
+	sendLock := &sync.Mutex{}
+	return func(msg *proto.PullMessageResponse) {
+		m := pubsub.NewMessage{
+			Data:        msg.Data,
+			ContentType: &msg.ContentType,
+			Topic:       msg.Topic,
+			Metadata:    msg.Metadata,
+		}
+		var ackError *proto.AckMessageError
+
+		if err := handler(ctx, &m); err != nil {
+			p.logger.Errorf("error when handling message on topic %s", msg.Topic)
+			ackError = &proto.AckMessageError{
+				Message: err.Error(),
+			}
+		}
+
+		// As per documentation:
+		// When using streams,
+		// one must take care to avoid calling either SendMsg or RecvMsg multiple times against the same Stream from different goroutines.
+		// In other words, it's safe to have a goroutine calling SendMsg and another goroutine calling RecvMsg on the same stream at the same time.
+		// But it is not safe to call SendMsg on the same stream in different goroutines, or to call RecvMsg on the same stream in different goroutines.
+		// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams
+		sendLock.Lock()
+		if err := streamingPull.Send(&proto.PullMessagesRequest{
+			AckMessageId: msg.Id,
+			AckError:     ackError,
+		}); err != nil {
+			p.logger.Errorf("error when ack'ing message %s from topic %s", msg.Id, msg.Topic)
+		}
+		sendLock.Unlock()
```

Comment:
As you start with go recently I would advise you to use defer with caution, especially in code that can be called many times without return - in a loop for instance. When you call "defer", there's no magic involved, the execution will be added on the top of execution stack which is... a stack, and it has its limits regard to size and memory usage

Having that said, stream.Send() is a blocking call that can take some time until returns, but on the other hand... each single call to handle() starts a new stack execution so the "loop" problem won't happen here, also, assuming that send() might panics (this is not fair assumption though, panic are not like an exception in other languages they are indeed to break programs - stop to run, people often recover from panics but that's another conversation) for an unexpected reason, so deferring the unlock is also good to prevent locking forever

Sometimes I find cleaner to not use defer when it involves only one line of code and this line is the return/end of function - seems overengineering

I'm going to change this, thanks